### ISSUE TITLE
Allow to rerun only failed tests

### ISSUE THREAD

**Issue by sylvain-hamel** [cypress-io/cypress#4886]
### Current behavior:

In the test runner, there is no option to only rerun failed tests. After investigating and fixing a failed test, I'd like the watcher to only rerun failed tests.

### Desired behavior:

Add a rerun "Rerun failed tests only" checkbox to the tests runner UI.

**Comment by jennifer-shehane** [comment]
Some similiar requests:

- https://github.com/cypress-io/cypress/issues/1313
- https://github.com/cypress-io/cypress/issues/1599

**Comment by Superkunair** [comment]
There's a plugin that can execute your failed test
[cypress-plugin-retries](https://github.com/Bkucera/cypress-plugin-retries)

**Comment by sylvain-hamel** [comment]
@jennifer-shehane and @Superkunair you both refer to issues and solutions to handle flaky tests by rerunning them. That's not what I'm looking for. I'm looking for a feature in the Test Runner UI. 

The use case is:

- I change some code and then run the tests using the Test Runner UI.
- Running them takes 10 minutes and 3 tests fail.
- For the next several minutes I'll focus on these 3 tests, changing code and rerunning them many times.
- While I do this, I want the Test Runner UI to only watch/rerun these 3 tests, until they pass.
- When I'm done fixing the issue, I can rerun the whole batch before I commit and push.

So this request is really for an improved developer experience while investigating issues, not as a solution to handle flaky tests.

**Comment by brian-mann** [comment]
What you're referring to is what we call burn-down mode internally.

**Comment by jennifer-shehane** [comment]
Yes, just tagging similar requests - this is a new kind of request which is why the issue was left open.

**Comment by DavidGWheeler** [comment]
> There's a plugin that can execute your failed test
> [cypress-plugin-retries](https://github.com/Bkucera/cypress-plugin-retries)

That is for re-running during execution to account for network issues?  We want a button in the runner marked "re-run failed tests only) to allow for debugging.  Don't confuse the issue.

**Comment by DavidGWheeler** [comment]
> @jennifer-shehane and @Superkunair you both refer to issues and solutions to handle flaky tests by rerunning them. That's not what I'm looking for. I'm looking for a feature in the Test Runner UI.
> 
> The use case is:
> 
> * I change some code and then run the tests using the Test Runner UI.
> * Running them takes 10 minutes and 3 tests fail.
> * For the next several minutes I'll focus on these 3 tests, changing code and rerunning them many times.
> * While I do this, I want the Test Runner UI to only watch/rerun these 3 tests, until they pass.
> * When I'm done fixing the issue, I can rerun the whole batch before I commit and push.
> 
> So this request is really for an improved developer experience while investigating issues, not as a solution to handle flaky tests.

THIS!  It is a really super simple and basic feature.  Every IDE has it.  Only Cypress runner fails to provide such a simple thing!

**Comment by DavidGWheeler** [comment]
> Some similiar requests:
> 
> * #1313
> * #1599

Neither of those issues relates to this issue in any way.

**Comment by Honigbaum** [comment]
I can agree with @DavidGWheeler, we're looking forward to see this feature too.

**Comment by juni0r** [comment]
I agree this would be an immensely helpful feature.

**Comment by samtsai** [comment]
I've found two options:
* Use `it.only(...)` in your tests
* https://github.com/bahmutov/cypress-skip-and-only-ui
![image](https://user-images.githubusercontent.com/225526/74962947-df27e400-53de-11ea-864b-c4581e54f2a7.png)

**Comment by LieAlbertTriAdrian** [comment]
This will be super helpful

**Comment by aesfer** [comment]
This is a must for test-driven development.

For me jest in watch mode (via vue-cli) has the ideal behaviour, it goes one step further and only re-runs the tests affected by code changes since the last time they ran (so that if there are 5 failing tests and I modify 1 of them only that one will re-run).

However I don't want to confuse the issue and as the OP suggested, an option for only re-running failing tests is a perfect first step.

**Comment by djaiss** [comment]
While I don’t agree with @DavidGWheeler 's statement about it being a feature that is both simple and easy (if it’s so easy, please submit a PR), I think this feature is really essential to the workflows on projects I’m working on.

I was really excited about the Cypress 5.0 when I read about the retry ability - but this is really not what I had in mind. I think having the ability to re-run a single failed test, or multiple failed tests in a row, would really help. 

On a test suite of 130 tests that take 20 minutes to run, seeing 13 failed tests (completely separated) and re-running them is a nightmare once I changed the behavior to fix them.

**Comment by narze** [comment]
In some other testing frameworks I've used they have this feature built-in (RSpec, Cucumber)

When the test fails it will create a file with filenames with line of code number of failing tests, and the next test run will focus only these tests. You don't have to "focus" the test at all. When all tests are passing, the file will be deleted.

**Comment by edimitchel** [comment]
Hi ! As I filled in #17803, it would nice to have TDD mode : GREEN, RED, REFACTOR which will make Cypress behaves differently depending to the mode.

**Comment by RyanAtViceSoftware** [comment]
Is there any update on this issue? I've seen a number of tickets related to this but it's not clear if this functionality is going to be added or where to track the progress.

**Comment by parkerault** [comment]
@jennifer-shehane Sorry to "me too" a feature request, but I think jest has made this an expected feature in a test runner (and much more important for integration test suites that can take multiple minutes to run in the average case scenario). Since it has been open for two years and other tickets are being closed as duplicates, is this currently on the roadmap?

**Comment by m3taphysics** [comment]
An update on this or a solution - it is soul destroying having to re-watch the entire set of tests to get to the failed ones for each iteration whereby you are trying to fix the problem

**Comment by edimitchel** [comment]
Please, bring the HMR test rerun!

**Comment by distante** [comment]
For people that center their work on TDD, this feature would be really helpful.

**Comment by baptisteArno** [comment]
> An update on this or a solution - it is soul destroying having to re-watch the entire set of tests to get to the failed ones for each iteration whereby you are trying to fix the problem

Agreed, this is extremely painful for a TDD workflow.

**Comment by Akuma-U1** [comment]
Any update on this feature?

**Comment by adamvanaken** [comment]
This request should not be limited to "failed" tests. While iteratively writing a test, it is a must to be able to tweak a single test then rerun it, whether it failed or not.

**Comment by danhooper** [comment]
@adamvanaken For that behavior you can add `it.only` or `describe.only` to focus on a single test.

**Comment by Carniatto** [comment]
@danhooper that's fine, but when running several specs (not possible on cypress 10 anymore), it would be very helpful to re-run only the failed ones. This way you shorten your list for each test fixed until you have all fixed while avoiding rerunning tests that already passed

**Comment by danhooper** [comment]
@Carniatto Agreed. My response was only to help with when you are developing tests. 

I very much want the ability to rerun failed tests.

**Comment by SwardCoder** [comment]
> 

did you find any solution for this brother, even I am looking for this 
as cypress test cases are independent of each other and the data being used by them is independent so yes even we are with the same requirement now to only rerun failed test cases
(but not retries to test flaky one)

**Comment by SwardCoder** [comment]
@jennifer-shehane any update on this its nearly 3 years still its open

**Comment by bahmutov** [comment]
FYI: https://glebbahmutov.com/blog/run-failed-tests/

**Comment by SwardCoder** [comment]
@bahmutov  wow this was very useful this opens a way to further investigate on how to integrate this in CI CD thank you :)

**Comment by metalocator** [comment]
> FYI: https://glebbahmutov.com/blog/run-failed-tests/

This is awesome.

**Comment by SwardCoder** [comment]
I Hope this solution will get integrated into command too so that we can use it via CI CD

**Comment by laerteneto** [comment]
I have a scenario in which I need to run my tests, then I need to restore the database, and finally run it again but only the failed ones from the previous run.

I was thinking in something like a file that is generated after a “cypress run”, so next time we run “cypress run” again, Cypress will first check if such file is already generated. If so, Cypress will run only the test cases listed in this specific file. 

I took some time investigating some options on the internet, and I don’t see exactly something that fits it. I know this PR is about the Test Runner over the UI, but it would be great to have it like I described for a CI CD approach.

**Comment by vedmant** [comment]
This is quite important functionally that Cypress lacks, and it's still only in proposal stage for 5 years. 
https://github.com/bahmutov/cypress-skip-and-only-ui would a be solution, but unfortunately it's not maintained.

**Comment by svetlana-napris-elucid** [comment]
Our team also suffering without the solution. I would say we need it everywhere: in CI/CD, while running in a headless mode locally as well as through UI.
It is really important feature, please take it into consideration!

**Comment by dennisbergevin** [comment]
Feel free to take a look at this plugin, which is companion to @bahmutov/cy-grep: 

https://github.com/dennisbergevin/cypress-plugin-last-failed

**Comment by SwardCoder** [comment]
Hi there
I tried to use it in my typescript based Cypress project and it's not
working for me

It's showing errors like

**Comment by FrenchMajesty** [comment]
In 2024, this would **still** be a massively valuable feature while working on tests.

**Comment by bcsgh** [comment]
Heck, I'd be happy if there was even just a "Rerun Scenario" button next to each scenario. 

Playing around with tags and the like seems kind of pointless given that I wouldn't be doing stuff in the UI if I wanted to be controlling stuff from outside the UI.
